### DISCUSSION TITLE
Dynamic keys

### DISCUSSION THREAD

**Discussion by roopakv** [47ng/nuqs#565] [Ideas] [answered]
We have an interesting case where we can have dynamic keys which vary per tenant of our application. These keys end up with just a string as a a value. Today there is no way to define all `query params that start with 'custom.': parseAsString`. 

Ideally being able to do that and return that as an array would be very neat. 

Our current approach is JSON, or we also played around with a custom serializer de-serializer.

Curious if there is a way to solve it today that we didnt think of.

**Comment by franky47** [comment, marked as answer]
This is what `useQueryStates` (plural) was intended for: allowing for arbitrary sets of keys. Here, you could calculate the list of custom keys based on some logic, then feed it to `useQueryStates`:

```tsx
'use client'

import { useSearchParams } from 'next/navigation'
import { parseAsString, useQueryStates } from 'nuqs'
import { Suspense } from 'react'

export default function Page() {
  return (
    <Suspense>
      <Client />
    </Suspense>
  )
}

function Client() {
  const sp = useSearchParams()
  const parsers = Object.fromEntries(
    Array.from(sp?.keys() ?? [])
      .filter(key => key.startsWith('custom.'))
      .map(key => [key, parseAsString.withDefault('')])
  )
  const [values, set] = useQueryStates(parsers)
  return (
    <ul>
      {Object.entries(values).map(([key, val]) => (
        <li key={key}>
          <span>
            {key}: {val}
          </span>
          <button onClick={() => set({ [key]: val + val.length })}>Set</button>
        </li>
      ))}
    </ul>
  )
}
```

https://github.com/47ng/nuqs/assets/1174092/f0bb586e-924b-47e8-a7e4-037b39ed1e4c

**Reply by roopakv** [reply]
ahh yeah we are using useQueryStates, but i didnt think of constructing that object dynamically! thank you for the quick response! let me give it a shot

**Reply by jagoral** [reply]
@franky47 looks good :) any ideas how to use it on server side with `createSearchParamsCache`?

**Reply by franky47** [reply]
@jagoral Same idea, using the search params coming from the page props. However you have to explicitly type the parsers object with keys of `custom.${string}` to have some type-safety on the prefix when accessing keys:

```tsx
import {
  createSearchParamsCache,
  parseAsString,
  type ParserBuilder,
  type SearchParams
} from 'nuqs/server'
import { Suspense } from 'react'

export default function Page({ searchParams }: { searchParams: SearchParams }) {
  const parsers: Record<
    `custom.${string}`,
    ParserBuilder<string>
  > = Object.fromEntries(
    Array.from(Object.keys(searchParams) ?? [])
      .filter(key => key.startsWith('custom.'))
      .map(key => [key, parseAsString.withDefault('')])
  )
  const id = 'foo'
  const cache = createSearchParamsCache(parsers)
  cache.parse(searchParams)
  const foo = cache.get(`custom.${id}`) // this will fail if trying to access 'bar.whatever', invalid prefix
  return (
    <Suspense>
      <Client />
    </Suspense>
  )
}
```

Edit: wait no this won't work actually. If you try and access a param that's not there (eg: hardcoding `custom.${foo}`), it will throw an error ([`NUQS-500`](https://github.com/47ng/nuqs/blob/next/errors/NUQS-500.md)).

In normal cases where you define your params without ambiguity, this is not an issue and an empty cache value is usually a good indication that `parse` hadn't been called, but here since keys can be ambiguous, this message is actually misleading.

**Comment by mjewell** [comment]
It seems like this doesn't allow you to set unknown keys. I have a use case where a table has some filters but we don't necessarily know what at the time we define the state (we're using tanstack table and want to sync the columnState to the url). If a key isn't already in the URL and you try to set it using this approach it just ignores that value because [it doesn't find a parser](https://github.com/47ng/nuqs/blob/next/packages/nuqs/src/useQueryStates.ts#L260).

Is there any way to support that? Specifically what I was hoping for is something where I can do

```tsx
  // starting url is `?custom.a=a` or just empty
  const sp = useSearchParams()
  const parsers = Object.fromEntries(
    Array.from(sp?.keys() ?? [])
      .filter(key => key.startsWith('custom.'))
      .map(key => [key, parseAsString.withDefault('')])
  )
  // parsers becomes `{ a: parseAsString.withDefault('') }` or `{}`
  const [values, set] = useQueryStates(parsers)

  // I want to then be able to do this, but it doesn't work because 'custom.b' isn't in the url.
  set({ 'custom.a': 'a', 'custom.b': 'b' });
```

**Reply by franky47** [reply]
Indeed that wouldn't work, because nuqs would have no idea how to serialise that state when setting it.

If you have a way to list those a, b etc values for the key names (eg: from the table state), then you could build the parser object from that.

As for the `custom.` prefix, this is something that could be done with urlKeys:

https://nuqs.47ng.com/docs/batching#shorter-search-params-keys

**Reply by mjewell** [reply]
Makes sense, thanks for the quick response. It's not really easy for me to know the keys up front so it'd be great if this was supported somehow, e.g. if we could provide a function that returned a parser based on the value that was being written. For now I am using a single useQueryState and parsing/serializing the keys into a special format within that key which is fine but just looks a bit awkward
